### PR TITLE
Feat: Credential Service

### PR DESCRIPTION
**Pull request by techsavvyash**
## Description 

This PR contains the actual code for the Credential service unlike the previous PR which was submodules based.

- [Postman Collection](https://www.postman.com/nksss-dev/workspace/ulp/collection/18245860-091875ae-0cd1-41c7-877d-725941ba561f?action=share&creator=18245860)
- [API Compliance Gist](https://gist.github.com/techsavvyash/923b8afabf472da57afcc948cac2cf4b)

### CODE REVIEW

File: services/credentials-service/prisma/migrations/20230304094253_migrate_mongodb_to_psql/migration.sql [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    "unsigned" JSONB NOT NULL,
+    "signed" JSONB,
+    "created_at" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    "updated_at" TIMESTAMP(3) DEFAULT CURRENT_TIMESTAMP,
+
+    CONSTRAINT "VC_pkey" PRIMARY KEY ("id")
+);
+
+-- CreateTable
+CREATE TABLE "VCV2" (
```
[the remark this answers is not in the document]
Comment (techsavvyash) [reply services/credentials-service/prisma/migrations/20230304094253_migrate_mongodb_to_psql/migration.sql:21]: Since the APIs are not connected with the registry right now, so we have a postgres instance associated wih the credential service to save the issued credentials in the postgres instance. This table stores those credentials.

File: services/credentials-service/prisma/migrations/20230304094253_migrate_mongodb_to_psql/migration.sql [hunk 2]
```diff
@@ -0,0 +1,68 @@
+-- CreateEnum
+CREATE TYPE "VCStatus" AS ENUM ('PENDING', 'ISSUED', 'REVOKED');
+
+-- CreateTable
+CREATE TABLE "VC" (
```
[the remark this answers is not in the document]
Comment (techsavvyash) [reply services/credentials-service/prisma/migrations/20230304094253_migrate_mongodb_to_psql/migration.sql:5]: This table has been deprecated and will be removed. Initially it was meant to store the issued credentials in the postgres db associated with the credentials service.

File: services/credentials-service/prisma/migrations/20230304094253_migrate_mongodb_to_psql/migration.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (tejash-jl) [reply services/credentials-service/prisma/migrations/20230304094253_migrate_mongodb_to_psql/migration.sql:21]: we need to fix the table name

**Comment by techsavvyash** [comment]
Hey @tejash-jl,
If everything is aligned can we merge this PR and #218 and #219 ?
Thanks

cc: @chakshugautam

File: services/credentials-service/prisma/migrations/20230304094253_migrate_mongodb_to_psql/migration.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (tejash-jl) [reply services/credentials-service/prisma/migrations/20230304094253_migrate_mongodb_to_psql/migration.sql:21]: are these fixed? @techsavvyash

**Comment by tejash-jl** [comment]
remove commented codes

**Comment by tejash-jl** [comment]
integrate e2e test to the top level Makefile

**Comment by tejash-jl** [comment]
Create e2e tests for all v2 APIs
Existing tests are present here, https://github.com/Sunbird-RC/sunbird-rc-core/tree/main/java/apitest/src/test/java/e2e/registry

File: Makefile [hunk 3]
```diff
@@ -41,16 +43,16 @@ test: build
 	@echo "Starting the test" && sh build/wait_for_port.sh 8081
 	@docker-compose ps
 	@curl -v http://localhost:8081/health
-	@cd java/apitest && ../mvnw -Pe2e test || echo 'Tests failed'
+	@cd java/apitest && ../mvnw -Pe2e test
```
Comment (srprasanna) [inline Makefile:46]: Could you add the tests failure message as well.

File: java/apitest/src/test/java/e2e/registry/registry.feature [hunk 4]
```diff
@@ -223,7 +223,7 @@ Feature: Registry api tests
     When method post
     Then status 200
     And response.result.Teacher.transactionId.length > 0
-    * sleep(3000)
+    * sleep(7000)
```
Comment (srprasanna) [inline java/apitest/src/test/java/e2e/registry/registry.feature:226]: Why is this changed.

File: java/apitest/src/test/java/e2e/registry/registry.feature [hunk 5]
```diff
@@ -468,10 +468,10 @@ Feature: Registry api tests
     And response.params.errmsg == "Schema 'Teacher1' not found"
  # patch unavailable schema with sign
     Given url registryUrl
-    And path '/api/v1/Teacher/sign'
+    And path '/api/v1/Teacher1/sign'
```
Comment (srprasanna) [inline java/apitest/src/test/java/e2e/registry/registry.feature:471]: Why was this changed.

File: java/apitest/src/test/java/e2e/registry/registry.feature [hunk 6]
```diff
@@ -468,10 +468,10 @@ Feature: Registry api tests
...
     And header Authorization = admin_token
     And request read('TeacherRequest.json')
-    When method patch
+    When method get
```
Comment (srprasanna) [inline java/apitest/src/test/java/e2e/registry/registry.feature:474]: Why is this changed.

File: services/credentials-service/src/credentials/credentials.service.spec.ts [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        })
+      ).toBeInstanceOf(Array);
+    } catch (e) {
+      expect(e.message).toBe(
+        'No credentials found for the given subject or issuer'
+      );
+    }
+  });
+
+  // it('should show a stremable file', async () => {
```
Comment (srprasanna) [inline services/credentials-service/src/credentials/credentials.service.spec.ts:251]: Please remove comments.

File: services/credentials-service/src/credentials/dto/getCredentialById.dto.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { ApiProperty } from "@nestjs/swagger";
+import { W3CCredential } from "did-jwt-vc";
+import { CredentialStatus, DateType, IssuerType } from "did-jwt-vc/lib/types";
+import { Extensible } from "did-resolver";
+
+
+// class Credential implements W3CCredential {
```
Comment (srprasanna) [inline services/credentials-service/src/credentials/dto/getCredentialById.dto.ts:7]: Remove commented Code.

File: java/apitest/src/test/java/e2e/registry/registry.feature (hunk 4 above)
Comment (tushar5526) [reply java/apitest/src/test/java/e2e/registry/registry.feature]: This happened due to one of the rebase probably, undid all these changes.

File: java/apitest/src/test/java/e2e/registry/registry.feature (hunk 5 above)
Comment (tushar5526) [reply java/apitest/src/test/java/e2e/registry/registry.feature]: resolved to original

File: java/apitest/src/test/java/e2e/registry/registry.feature (hunk 6 above)
Comment (tushar5526) [reply java/apitest/src/test/java/e2e/registry/registry.feature]: reverted to original

File: services/credentials-service/src/credentials/dto/getCredentialById.dto.ts (hunk 8 above)
Comment (tushar5526) [reply services/credentials-service/src/credentials/dto/getCredentialById.dto.ts]: done, thanks

File: services/credentials-service/src/credentials/credentials.service.spec.ts (hunk 7 above)
Comment (tushar5526) [reply services/credentials-service/src/credentials/credentials.service.spec.ts]: done, thanks

**Comment by tushar5526** [comment]
Screenshot 2023-07-11 at 7 26 42 PM

File: services/identity-service/docker-compose.yml [hunk 9]
```diff
@@ -50,7 +48,7 @@ services:
       VAULT_ADDR: "http://vault:8200"
       # This will be replaced automatically on initialisation 
       # make compose-init will call setup_vault.sh
-      VAULT_TOKEN: hvs.ZBCv64FQYZibBmfa6cgcc8ls
+      VAULT_TOKEN: hvs.ANn8uBxHYvRCv5Q3FXuqzoIG
```
Comment (srprasanna) [inline services/identity-service/docker-compose.yml:51]: Can you move these to .env file.

**Review by srprasanna — APPROVED** [review]

**Comment by srprasanna** [comment]
> Screenshot 2023-07-11 at 7 26 42 PM

Can you try to increase overall coverage. Some of the files have about 40-50% coverage. Can you look at the credential service and controller.

**Comment by tushar5526** [comment]
Screenshot 2023-07-18 at 11 43 41 AM

I have increased the test coverage. Please take a look. `72-101` line in the controller is tested in `e2e` tests but not reflected here. It is the header parsing logic for `accept` header

**Comment by tushar5526** [comment]
Moving to vault token to `.env` will require a bit of refactoring in all 3 services and I am planning to make it as a different PR. For now, I am just removing it from the compose.

**Review by srprasanna — APPROVED** [review]
